Add secure mount function

leonjia0112 opened the pull request:
1. contains unsafe libc function

leonjia0112 commented:
Added secure mount function in a separated secure mount file. Test passed, ready for a review.

frozencemetery commented on src/crypto.rs:
```diff
@@ -42,6 +42,7 @@ pub fn do_hmac(
  * PBKDF2 function defaults to SHA-1 unless otherwise specified, and
  * Python-Keylime uses this default.
  */
+#[allow(dead_code)]
```
Please don't add these.

leonjia0112 replied:
Without this would cause function unused warning if that is not an issue.

leonjia0112 replied:
Got it. Thanks.

frozencemetery commented on src/common.rs:
```diff
@@ -20,6 +20,15 @@ pub static IMA_ML_STUB: &'static str =
 pub static IMA_ML: &'static str =
     "/sys/kernel/security/ima/ascii_runtime_measurements";
 pub static KEY: &'static str = "secret";
+pub static MOUNT_SECURE: bool = true;
```
This shouldn't be a global constant.

leonjia0112 replied:
Moved to local constant.

frozencemetery commented on src/common.rs:
```diff
@@ -99,6 +108,24 @@ pub fn get_restful_parameters(urlstring: &str) -> HashMap<&str, &str> {
+    unsafe {
+        let euid = libc::geteuid();
+        if euid == 0 {
+            libc::chown(path.as_bytes().as_ptr() as *const i8, 0, 0);
```
Handle return code on this.

leonjia0112 replied:
A status code is returned to caller with execution result log.

leonjia0112 commented:
Test passed, ready for a review. Thanks.

leonjia0112 commented on src/common.rs:
```diff
@@ -99,6 +107,42 @@ pub fn get_restful_parameters(urlstring: &str) -> HashMap<&str, &str> {
+ * execution result.
+ */
+pub fn chownroot(path: String) -> Option<i32> {
+    unsafe {
```
Thanks for your review. 

I changed the error handling in chownroot() function by return a result<> that could to be handled by the caller. The mount() function is also changed to return a result<> for error handling. Also, early return is implemented. Err is returned for error result.

The chownroot() function is actually used in multiple places in keylime_node, including functions in Common.rs and secure_moutn.rs.

Secure_mount is used for checking the mount status of the secure directory. It makes sure there is a tmpfs for keylime node to store keys, which is /var/lib/keylime/secure/ directory normally.

leonjia0112 commented:
Changed based on the request. Test passed, and ready for another review. Thanks.

I changed the commit message and added some inline comments for the code. Hope these could make it clearer about what the change is used for.

frozencemetery commented on src/common.rs:
```diff
@@ -99,6 +108,33 @@ pub fn get_restful_parameters(urlstring: &str) -> HashMap<&str, &str> {
+        if libc::geteuid() != 0 {
+            error!("Privilege level unable to change ownership to root for file: {}", path);
+            return Err(-1);
+        } else {
```
Close.  Use more early exit  The flow should be like:

```rust
if getuid() != 0 {
    return;
}
if chown() != 0 {
    return;
}
Ok()
```

This keeps the branching manageable, which is easier to read and reason about.

leonjia0112 commented:
Implemented request changes. Test passed. Ready for a review. Thanks.

frozencemetery commented:
Change merged.  Please take a look at the rewritten commit message.
